Add dotnet-aspnetcore REST API authoring skills (controllers + Minimal APIs) (#848)
dotnet/skills: plugins/dotnet-aspnetcore/skills/structured-logging/SKILL.md

---
name: structured-logging
description: >
  Write diagnosable structured logs in ASP.NET Core: message templates with named placeholders, a scope
  to correlate an operation's lines, appropriate levels, and no sensitive data.
  USE FOR: adding logging to an operation so it can be searched and correlated in production; using
  ILogger message templates with named properties instead of interpolated strings; correlating log lines
  for one request/operation with BeginScope or a correlation id; choosing log levels; logging exceptions;
  keeping secrets and PII out of logs; LoggerMessage source generation on hot paths.
  DO NOT USE FOR: distributed tracing/metrics wiring (that is OpenTelemetry); global exception-to-response
  mapping (use IExceptionHandler); audit persistence in the database.
license: MIT
---

# Structured Logging

> 

Reviewer: ```suggestion

For controller operations, inject `ILogger<ControllerType>` and treat one action as one operation. Open a
scope after the request has been bound and the resource identity is known, then log the operation and its
outcome inside that scope. Use identifiers already present in the request or entity; do not invent a tenant,
user, or correlation field that the model does not provide.

```

Reviewer: ```csharp
[ApiController]
[Route("api/namespaces")]
public sealed class NamespacesController(
  MessagingDbContext db,
  ILogger<NamespacesController> logger) : ControllerBase
{
  [HttpDelete("{name}")]
  public async Task<IActionResult> Delete(string name, CancellationToken ct)
  {
    var operationId = HttpContext.TraceIdentifier;
    using var scope = logger.BeginScope(new Dictionary<string, object?>
    {
      ["Operation"] = "DeleteNamespace",
      ["NamespaceName"] = name,
      ["OperationId"] = operationId
    });

    var entity = await db.Namespaces.SingleOrDefaultAsync(
      x => x.Name == name && !x.IsDeleted, ct);
    if (entity is null)
    {
      logger.LogWarning("Namespace not found for deletion");
      return NotFound();
    }

    entity.IsDeleted = true;
    await db.SaveChangesAsync(ct);
    logger.LogInformation("Namespace deleted");
    return NoContent();
  }
}
```

Apply the same shape to create and update: log a normal-operation event after a successful save, and log a
warning for expected misses or conflicts. Log unexpected exceptions at `Error` with the exception object and
rethrow or let the application's exception handler translate them. Do not log a full request body, tags,
connection strings, access tokens, or exception messages as template arguments.

## Message templates with named placeholders, not interpolation

Pass the values as arguments to a template. Each `{Name}` becomes a structured property the log store can index and query; interpolation throws that away and leaves only flat text.

```csharp
logger.LogInformation("Created order {OrderId} for customer {CustomerId}", order.Id, customerId);
```

❌ `logger.LogInformation($"Created order {order.Id} for customer {customerId}");` - interpolated: no `OrderId`/`CustomerId` properties, just a string you cannot filter on.

## Correlate an operation's lines with a scope

A scope attaches the same properties to every entry written inside it, so all lines for one request, order, or tenant can be retrieved together.

```csharp
using (logger.BeginScope(new Dictionary<string, object>
{
    ["OrderId"] = order.Id,
    ["TenantId"] = tenantId
}))
{
    logger.LogInformation("Validating order");
    logger.LogInformation("Charging payment");
    // both lines carry OrderId and TenantId
}
```

The framework already attaches request-level correlation (the trace identifier) to logs; add a scope for the identifiers that matter to *your* operation.

## Levels, exceptions, and content

> - Use levels by severity: `Information` for normal flow, `Warning` for recoverable problems, `Error` for failures. Do not log everything at one level.

Reviewer: ```suggestion
- Use levels by severity: `Information` for successful normal flow, `Warning` for expected misses or recoverable conflicts, and `Error` for unexpected failures. Do not log everything at one level.
```

- Log an exception by passing it as the **first argument**, not by string-formatting it: `logger.LogError(ex, "Failed to create order {OrderId}", order.Id);`.

> - Never log secrets, tokens, full PII, or request/response bodies. Log identifiers and outcomes, not payloads.

Reviewer: ```suggestion
- Put stable operation and resource identifiers in the scope or event template. A scope should not replace the event message: the event still says what happened and whether it succeeded.
- Never log secrets, tokens, full PII, or request/response bodies. Log identifiers and outcomes, not payloads.
```
